<a href="https://colab.research.google.com/github/viandre7/Viviana_GarciaMonje_KC/blob/main/Taller3_ok.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python en un contexto de Negocio de aerolineas (Caso practico)

**Contexto empresarial.** Es un empleado de GrowthAir, una compañía aérea comercial en crecimiento. En los últimos años, GrowthAir ha expandido los servicios de vuelos de lujo a ubicaciones en todo el mundo. Tras el excelente desempeño de su equipo en la identificación de nuevas oportunidades comerciales el año pasado, se le ha encomendado la tarea de identificar los principales países para expandir aún más el servicio de vuelos de lujo de GrowthAir.

**Problema empresarial.** Su gerente le ha pedido que responda a la siguiente pregunta: "¿En qué países debería GrowthAir expandir su servicio de vuelos de lujo?"

**Contexto analítico.** Los datos relevantes son una serie de estimaciones de éxito (es decir, probabilidades de éxito) que han elaborado sus equipos internos de investigación de mercados. Usando su capacidad para realizar análisis de datos en Python, se embarcará en resumir las estimaciones de éxito disponibles para producir una recomendación concisa para su jefe.

In [ ]:
#Datos sobre la probabilidad de éxito de la expansión por estimaciones de países
success_estimates = {'Australia': [0.6, 0.33, 0.11, 0.14],
                     'France': [0.66, 0.78, 0.98, 0.2],
                     'Italy': [0.6],
                     'Brazil': [0.22, 0.22, 0.43],
                     'USA': [0.2, 0.5, 0.3],
                     'England': [0.45],
                     'Canada': [0.25, 0.3],
                     'Argentina': [0.22],
                     'Greece': [0.45, 0.66, 0.75, 0.99, 0.15, 0.66],
                     'Morocco': [0.29],
                     'Tunisia': [0.68, 0.56],
                     'Egypt': [0.99],
                     'Jamaica': [0.61, 0.65, 0.71],
                     'Switzerland': [0.73, 0.86, 0.84, 0.51, 0.99],
                     'Germany': [0.45, 0.49, 0.36]}

In [ ]:
print(success_estimates)

{'Australia': [0.6, 0.33, 0.11, 0.14], 'France': [0.66, 0.78, 0.98, 0.2], 'Italy': [0.6], 'Brazil': [0.22, 0.22, 0.43], 'USA': [0.2, 0.5, 0.3], 'England': [0.45], 'Canada': [0.25, 0.3], 'Argentina': [0.22], 'Greece': [0.45, 0.66, 0.75, 0.99, 0.15, 0.66], 'Morocco': [0.29], 'Tunisia': [0.68, 0.56], 'Egypt': [0.99], 'Jamaica': [0.61, 0.65, 0.71], 'Switzerland': [0.73, 0.86, 0.84, 0.51, 0.99], 'Germany': [0.45, 0.49, 0.36]}


Observe el reordenamiento de los elementos del diccionario cuando imprimimos en comparación con el orden en el que originalmente definimos el diccionario. Este es un aspecto clave de los tipos de datos de diccionario: ¡están desordenados! (Esto es muy diferente en comparación con los tipos de datos de lista, que están ordenados.).

Ahora intuitivamente, nos gustaría recomendar que la empresa se esfuerce en el país con la estimación de éxito más alta. Pero, ¿qué significa esto cuando hay múltiples estimaciones de éxito para algunos países y solo una para otros? Exploraremos esto a continuación.


# Interactuemos con los diccionarios y listas

Si observa detenidamente el diccionario ```success_estimates```, verá que algunos países solo tienen una estimación de éxito, mientras que otros tienen muchas. Por ejemplo, Inglaterra solo tiene una estimación en su lista [0,45], mientras que Jamaica tiene tres estimaciones en su lista [0,61, 0,65, 0,71]. Acerquémonos a Jamaica y echemos un vistazo a algunas estadísticas resumidas de las estimaciones.

En Python, el tipo de diccionario tiene métodos incorporados (funciones, que discutiremos más adelante) para acceder a las claves y valores del diccionario. Estos métodos se llaman escribiendo  ```.keys()``` o ```.values ()``` después del objeto dictonary. Cambiaremos el tipo de retorno de las llamadas ```.keys()``` y ```.values()``` a una lista utilizando el método ```list()```.

In [ ]:
list(success_estimates.keys())

['Australia',
 'France',
 'Italy',
 'Brazil',
 'USA',
 'England',
 'Canada',
 'Argentina',
 'Greece',
 'Morocco',
 'Tunisia',
 'Egypt',
 'Jamaica',
 'Switzerland',
 'Germany']

In [ ]:
list(success_estimates.values())

[[0.6, 0.33, 0.11, 0.14],
 [0.66, 0.78, 0.98, 0.2],
 [0.6],
 [0.22, 0.22, 0.43],
 [0.2, 0.5, 0.3],
 [0.45],
 [0.25, 0.3],
 [0.22],
 [0.45, 0.66, 0.75, 0.99, 0.15, 0.66],
 [0.29],
 [0.68, 0.56],
 [0.99],
 [0.61, 0.65, 0.71],
 [0.73, 0.86, 0.84, 0.51, 0.99],
 [0.45, 0.49, 0.36]]

Haremos uso del acceso a las llaves primarias y valores de un diccionario más adelante en el caso al comparar las estimaciones de numerosos países. Por ahora, recuerde que puede acceder a la lista completa de llaves primarias o valores de un diccionario simplemente llamando a los métodos integrados.

# Ejercicio 1
También nos gustaría comprobar si el nombre de un país es una de las claves del diccionario. Verificar si Morocco y Japan se encuentran en ``success_estimates``

De ser asi obtener la lista de valores para Morocco y Japan


In [ ]:
def ver_claves(pais):

  if pais in success_estimates:
    result= print('La lista de los valores de',pais,'es: ',success_estimates[pais])
    return result

  else:
    result=print('El pais', pais, 'no se encuentra')
    return result

In [ ]:
ver_claves(input('Por favor digite el nombre del país que desea verificar: '))

Por favor digite el nombre del país que desea verificar: Morocco
La lista de los valores de Morocco es:  [0.29]


# Ejercicio 2
Imprimir la longitud de las listas de estimaciones de éxito para Francia, Grecia y Marruecos.


In [ ]:
def longitud(long_):
  if long_ in success_estimates:
    list_len = len(success_estimates[long_])
  return list_len

In [ ]:
longitud(input('Por favor digite el nombre del país: '))

Por favor digite el nombre del país: France


4

# Ejercicio 3

¿Cuál de las siguientes opciones sería útil para almacenar estimaciones de éxito del proyecto si estuvieran disponibles a nivel regional en lugar de a nivel nacional?

a) Lista

(b) Diccionario

(c) Flotantes

(d) Cadena

Diccionario, esto teniendo en cuenta las propiedades de los mismos, las cuales nos permiten definir y acceder a los compnentes mediante llaves/claves, así como de los mismos valores, segun sea lo solicitado.son de acceso rápido.

Adicional para el presente proyecto, sin importar que los datos o estimaciones de exito sean a nivel nacional o regional, el beneficio de los diccionarios es el mismo

## Calcular una estimación de éxito promedio específica de un país

Continuando con nuestro análisis sobre Jamaica, la lista contiene tres números, [0,61, 0,65, 0,71]. Recuerde que estos números son del tipo ```float``` en Python, que almacena valores decimales numéricos. Una forma lógica de resumir estas estimaciones para poder compararlas entre países es utilizar el promedio aritmético. Usemos operadores artiméticos básicos para calcular la estimación de éxito promedio para Jamaica, almacenando el resultado en una nueva variable ```avg_jamaica```:

In [ ]:
avg_jamaica = (0.61 + 0.65 + 0.71) / 3
print(avg_jamaica)

0.6566666666666666


Vemos que la probabilidad promedio de éxito estimada para Jamaica es de aproximadamente 0.657. Sin embargo, producimos esta estimación codificando manualmente los valores. Si hiciéramos esto para todos los países, llevaría bastante tiempo. Por eso, nos gustaría utilizar una forma más automatizada de producir el promedio.

Para producir un promedio, podemos utilizar una <b>función</b>. Las funciones operan sobre datos y variables en Python para realizar una acción deseada. Las funciones pueden tener tanto <b> entradas </b> como <b> salidas </b>, al igual que los operadores matemáticos familiares como suma, resta, multiplicación y división (que tienen dos entradas y una salida). Si bien las funciones en Python aún pueden tener un propósito matemático, como elevar al cuadrado un número entero, Python permite un comportamiento de función más abstracto, como imprimir en la pantalla. En este caso, la función ```print()``` imprimirá su entrada en la pantalla.

In [ ]:
country_name = 'Jamaica'
jamaica_list = success_estimates[country_name] # lista de estimacione spara jamaica
print(jamaica_list)

[0.61, 0.65, 0.71]


In [ ]:
avg_jamaica = sum(jamaica_list) / len(jamaica_list)
min_jamaica = min(jamaica_list)
max_jamaica = max(jamaica_list)
print("Pais:",country_name,", Average:",avg_jamaica)
print("Pais:",country_name,", Min:",min_jamaica)
print("Pais:",country_name,", Max:",max_jamaica)

Pais: Jamaica , Average: 0.6566666666666666
Pais: Jamaica , Min: 0.61
Pais: Jamaica , Max: 0.71


In [ ]:
# Ahora redondeando
avg_jamaica = round(sum(jamaica_list) / len(jamaica_list),2)
min_jamaica = round(min(jamaica_list),2)
max_jamaica = round(max(jamaica_list),2)
print("Pais:",country_name,", Average:",avg_jamaica)
print("Pais:",country_name,", Min:",min_jamaica)
print("Pais:",country_name,", Max:",max_jamaica)

Pais: Jamaica , Average: 0.66
Pais: Jamaica , Min: 0.61
Pais: Jamaica , Max: 0.71


Escribamos un código para calcular el éxito promedio de cada país. Muestra (usando print ()) la estimación de éxito promedio de cada país en la pantalla. Las declaraciones impresas deben mostrar cada país en una nueva línea, por ejemplo:

In [ ]:
print("Pais:",'France',", Average:",sum(success_estimates['France']) / len(success_estimates['France']))
print("Pais:",'Brazil',", Average:",sum(success_estimates['Brazil']) / len(success_estimates['Brazil']))
print("Pais:",'Argentina',", Average:",sum(success_estimates['Argentina']) / len(success_estimates['Argentina']))
print("Pais:",'Germany',", Average:",sum(success_estimates['Germany']) / len(success_estimates['Germany']))
print("Pais:",'Australia',", Average:",sum(success_estimates['Australia']) / len(success_estimates['Australia']))
print("Pais:",'Canada',", Average:",sum(success_estimates['Canada']) / len(success_estimates['Canada']))
print("Pais:",'Greece',", Average:",sum(success_estimates['Greece']) / len(success_estimates['Greece']))
print("Pais:",'USA',", Average:",sum(success_estimates['USA']) / len(success_estimates['USA']))
print("Pais:",'Switzerland',", Average:",sum(success_estimates['Switzerland']) / len(success_estimates['Switzerland']))
print("Pais:",'Tunisia',", Average:",sum(success_estimates['Tunisia']) /len(success_estimates['Tunisia']))
print("Pais:",'Italy',", Average:",sum(success_estimates['Italy']) / len(success_estimates['Italy']))
print("Pais:",'Egypt',", Average:",sum(success_estimates['Egypt']) / len(success_estimates['Egypt']))
print("Pais:",'Jamaica',", Average:",sum(success_estimates['Jamaica']) / len(success_estimates['Jamaica']))
print("Pais:",'Morocco',", Average:",sum(success_estimates['Morocco']) / len(success_estimates['Morocco']))
print("Pais:",'England',", Average:",sum(success_estimates['England']) / len(success_estimates['England']))

Pais: France , Average: 0.655
Pais: Brazil , Average: 0.29
Pais: Argentina , Average: 0.22
Pais: Germany , Average: 0.4333333333333333
Pais: Australia , Average: 0.29500000000000004
Pais: Canada , Average: 0.275
Pais: Greece , Average: 0.61
Pais: USA , Average: 0.3333333333333333
Pais: Switzerland , Average: 0.7859999999999999
Pais: Tunisia , Average: 0.6200000000000001
Pais: Italy , Average: 0.6
Pais: Egypt , Average: 0.99
Pais: Jamaica , Average: 0.6566666666666666
Pais: Morocco , Average: 0.29
Pais: England , Average: 0.45


## Determinar sistemáticamente la estimación de éxito promedio para todos los países

El objetivo final de este análisis es una recomendación sobre dónde se deben considerar las oportunidades de expansión global. Para llegar a una conclusión, sería ideal tener la probabilidad de éxito promedio para cada país.

Para lograr esto, usaremos un elemento de flujo de control en Python: el <b> bucle for </b>. El bucle ```for``` permite ejecutar las mismas declaraciones una y otra vez (es decir, en bucle). Esto ahorra una cantidad significativa de tiempo codificando tareas repetitivas y ayuda en la legibilidad del código. La estructura general de un bucle for es:

```python
for iterator_variable in some_sequence:
    statements(s)
```

The for loop iterates over ```some_sequence``` and performs ```statements(s)``` at each iteration. That is, at each iteration the ```iterator_variable```  is updated to the next value in ```some_sequence```. As a concrete example, consider the loop:

```python
for i in [1,2,3,4]:
    print(i*i)
```

Aquí, el bucle for se imprimirá en la pantalla cuatro veces; es decir, imprimirá ```1``` en la primera iteración del ciclo,```4``` en la segunda iteración,```9``` en la tercera y ```16``` en el cuarto. Por lo tanto, la instrucción de bucle for iterará sobre todos los elementos de la lista ```[1,2,3,4]```, y en cada iteración actualiza la variable iteradora ```i``` a la siguiente valor en la lista ```[1,2,3,4]```.

Usemos un ciclo for en los datos de nuestro país obteniendo una lista de todas las claves en ```success_estimates```:


In [ ]:
country_name_list = list(success_estimates.keys())
print(country_name_list)

['Australia', 'France', 'Italy', 'Brazil', 'USA', 'England', 'Canada', 'Argentina', 'Greece', 'Morocco', 'Tunisia', 'Egypt', 'Jamaica', 'Switzerland', 'Germany']


Aquí podemos recorrer todos los elementos en ```country_name_list```, extraemos el valor correspondiente de ```success_estimates``` (que será del tipo list), y posteriormente tomamos la media de la lista. La impresión detallada lo guiará a través de la ejecución del bucle for.

# Ejericicio 4

Crear un loop que obtenga e imprima el promedio de la tasa de exito de cada pais

In [ ]:
def promedio():
  for i in success_estimates:
      print("País:", i)
      print("Promedio: ",round(sum(success_estimates[i])/len(success_estimates[i]),2))
      print("---------")

In [ ]:
promedio()

País: Australia
Promedio:  0.3
---------
País: France
Promedio:  0.66
---------
País: Italy
Promedio:  0.6
---------
País: Brazil
Promedio:  0.29
---------
País: USA
Promedio:  0.33
---------
País: England
Promedio:  0.45
---------
País: Canada
Promedio:  0.28
---------
País: Argentina
Promedio:  0.22
---------
País: Greece
Promedio:  0.61
---------
País: Morocco
Promedio:  0.29
---------
País: Tunisia
Promedio:  0.62
---------
País: Egypt
Promedio:  0.99
---------
País: Jamaica
Promedio:  0.66
---------
País: Switzerland
Promedio:  0.79
---------
País: Germany
Promedio:  0.43
---------


# Ejercicio 5
Escribir un bucle for para calcular el mínimo y el máximo de la lista de estimaciones de éxito de cada país, imprimiendo cada uno de forma consecutiva como en el ejemplo de bucle for anterior.

In [ ]:
def Minmax():
  for i in success_estimates:
      print("País:", i)
      print("Minimo:",min(success_estimates[i]),'y Máximo',max(success_estimates[i]))
      print("---------")


In [ ]:
Minmax()

País: Australia
Minimo: 0.11 y Máximo 0.6
---------
País: France
Minimo: 0.2 y Máximo 0.98
---------
País: Italy
Minimo: 0.6 y Máximo 0.6
---------
País: Brazil
Minimo: 0.22 y Máximo 0.43
---------
País: USA
Minimo: 0.2 y Máximo 0.5
---------
País: England
Minimo: 0.45 y Máximo 0.45
---------
País: Canada
Minimo: 0.25 y Máximo 0.3
---------
País: Argentina
Minimo: 0.22 y Máximo 0.22
---------
País: Greece
Minimo: 0.15 y Máximo 0.99
---------
País: Morocco
Minimo: 0.29 y Máximo 0.29
---------
País: Tunisia
Minimo: 0.56 y Máximo 0.68
---------
País: Egypt
Minimo: 0.99 y Máximo 0.99
---------
País: Jamaica
Minimo: 0.61 y Máximo 0.71
---------
País: Switzerland
Minimo: 0.51 y Máximo 0.99
---------
País: Germany
Minimo: 0.36 y Máximo 0.49
---------


# Ejercicio 6
Escriba el código para determinar el país con el mayor rango de estimaciones de éxito (es decir, la mayor diferencia entre la estimación más pequeña y la más grande para un país).

In [ ]:
def may_rango():
  rango={}

  for i in success_estimates:
    rang = max(success_estimates[i])-min(success_estimates[i])
    rango[i]=rang
 
  max_key = max(rango, key=rango.get)
  print("El país con el mayor rango de estimaciones de éxito es:",max_key,"y, su valor es:",rango[max_key])

In [ ]:
may_rango()

El país con el mayor rango de estimaciones de éxito es: Greece y, su valor es: 0.84


## Uso de listas por comprensión para determinar la cantidad de estimaciones para cada país

En el futuro, estamos interesados en conocer la cantidad de estimaciones de éxito disponibles para cada país. Python ofrece una forma concisa de lograr este objetivo mediante el uso de <b> listas por comprensión </b>.

Las listas por comprensión le permiten a uno construir una lista de manera concisa. Echemos un vistazo a cómo funciona esto.

In [ ]:
key_name_list = [i for i in success_estimates] # loop sobre cada item  i en success_estimates y ponerlo en una lista
key_name_list

['Australia',
 'France',
 'Italy',
 'Brazil',
 'USA',
 'England',
 'Canada',
 'Argentina',
 'Greece',
 'Morocco',
 'Tunisia',
 'Egypt',
 'Jamaica',
 'Switzerland',
 'Germany']

Aquí vemos que hemos hecho un bucle sobre cada clave del diccionario success_estimates (por lo tanto, cada país) y extraído el nombre del país, todo en una línea de código. También podemos acceder a los valores de cada clave en success_estimates.

In [ ]:
value_name_list = [success_estimates[i] for i in success_estimates] # loop sobre cada item i ensuccess_estimates y poner success_estimates[i] en la lista
value_name_list

[[0.6, 0.33, 0.11, 0.14],
 [0.66, 0.78, 0.98, 0.2],
 [0.6],
 [0.22, 0.22, 0.43],
 [0.2, 0.5, 0.3],
 [0.45],
 [0.25, 0.3],
 [0.22],
 [0.45, 0.66, 0.75, 0.99, 0.15, 0.66],
 [0.29],
 [0.68, 0.56],
 [0.99],
 [0.61, 0.65, 0.71],
 [0.73, 0.86, 0.84, 0.51, 0.99],
 [0.45, 0.49, 0.36]]

En la lista por comprensión anterior, cada valor de ```i``` es un nombre de país y el valor se devuelve cuando se llama a ```success_estimates [i]```. Vemos que esto es una forma efectiva y concisa de escribir un bucle for que crea una lista.

Podemos usar esto para determinar rápidamente cuántas estimaciones de éxito están disponibles para cada país.

In [ ]:
[[i,len(success_estimates[i])] for i in success_estimates]

[['Australia', 4],
 ['France', 4],
 ['Italy', 1],
 ['Brazil', 3],
 ['USA', 3],
 ['England', 1],
 ['Canada', 2],
 ['Argentina', 1],
 ['Greece', 6],
 ['Morocco', 1],
 ['Tunisia', 2],
 ['Egypt', 1],
 ['Jamaica', 3],
 ['Switzerland', 5],
 ['Germany', 3]]

# Ejercicio 7
Utilizando listas por comprensión, escriba un script para crear una <b> lista de listas </b> llamada ```sum_squares_list```, donde cada elemento de la lista es una lista de dos elementos [nombre del país, valor]. El elemento de valor en la lista debe ser la suma de cuadrados de las estimaciones de éxito de ese país. Por ejemplo, un elemento de ```sum_squares_list``` debe ser para Jamaica, donde la lista de dos elementos es [Jamaica, 1.2987] (desde 1.2987 = 0.61^ 2 + 0.65^ 2 + 0.71^ 2).

In [178]:
sum_squares_list = [[i,round(sum([j**2 for j in success_estimates[i]]),2)] for i in success_estimates]
sum_squares_list

[['Australia', 0.5],
 ['France', 2.04],
 ['Italy', 0.36],
 ['Brazil', 0.28],
 ['USA', 0.38],
 ['England', 0.2],
 ['Canada', 0.15],
 ['Argentina', 0.05],
 ['Greece', 2.64],
 ['Morocco', 0.08],
 ['Tunisia', 0.78],
 ['Egypt', 0.98],
 ['Jamaica', 1.3],
 ['Switzerland', 3.22],
 ['Germany', 0.57]]

# Ejercicio 8
Nos gustaría determinar la distribución alrededor de la estimación de éxito promedio para cada país. Utilizando listas por comprensión, escriba un codigo que reste la estimación de éxito promedio para un país determinado de cada estimación de éxito para ese país. Almacene los resultados en una lista llamada ```remove_mean_list```. Redondea los valores a dos decimales. Su salida debe producir la siguiente lista de listas:


```
[['Australia', [0.3, 0.03, -0.19, -0.16]],
 ['France', [0.01, 0.12, 0.32, -0.46]],
 ['Italy', [0.0]],
 ['Brazil', [-0.07, -0.07, 0.14]],
 ['USA', [-0.13, 0.17, -0.03]],
 ['England', [0.0]],
 ['Canada', [-0.03, 0.02]],
 ['Argentina', [0.0]],
 ['Greece', [-0.16, 0.05, 0.14, 0.38, -0.46, 0.05]],
 ['Morocco', [0.0]],
 ['Tunisia', [0.06, -0.06]],
 ['Egypt', [0.0]],
 ['Jamaica', [-0.05, -0.01, 0.05]],
 ['Switzerland', [-0.06, 0.07, 0.05, -0.28, 0.2]],
 ['Germany', [0.02, 0.06, -0.07]]]
```

In [198]:
remove_mean_list = [[i,[round(j-(sum(success_estimates[i])/len(success_estimates[i])),2) for j in success_estimates[i]]] for i in success_estimates]
remove_mean_list


[['Australia', [0.3, 0.03, -0.19, -0.16]],
 ['France', [0.01, 0.12, 0.32, -0.46]],
 ['Italy', [0.0]],
 ['Brazil', [-0.07, -0.07, 0.14]],
 ['USA', [-0.13, 0.17, -0.03]],
 ['England', [0.0]],
 ['Canada', [-0.03, 0.02]],
 ['Argentina', [0.0]],
 ['Greece', [-0.16, 0.05, 0.14, 0.38, -0.46, 0.05]],
 ['Morocco', [0.0]],
 ['Tunisia', [0.06, -0.06]],
 ['Egypt', [0.0]],
 ['Jamaica', [-0.05, -0.01, 0.05]],
 ['Switzerland', [-0.06, 0.07, 0.05, -0.28, 0.2]],
 ['Germany', [0.02, 0.06, -0.07]]]

Según el análisis anterior, vemos que las estimaciones de éxito medio de los países varían ampliamente, desde el más bajo, Canadá = 0,275, hasta el más alto, Egipto = 0,99. Sin embargo, observe que la media de Egipto se calcula a partir de 1 estimación de éxito. ¿Confiamos en confiar en una única estimación como proxy de la estimación de éxito promedio?

Dado que el proyecto de expansión global utilizará valiosos recursos de la empresa, decidimos que es mejor restringir nuestro análisis a países que tienen dos o más estimaciones de éxito. Para realizar esta tarea, usaremos una estructura de control en Python conocida como la <b> instrucción if ... elif ... else </b>. Sigue la estructura general.


```python
if test_expression_1:
    block1_statement(s)
elif test_expression_2:
    block2_statement2(s)
else:
    block3_statement(s)
```

Aquí, ```test_expression_1``` y ```test_expression_2``` deben evaluarse como ```True``` o ```False```, un tipo <b>booleano</b> de Python. El tipo booleano está asociado con variables que son ```Verdadero``` o ```Falso```.

Si ```test_expression_1``` es True, ```block1_statement(s)``` se ejecutará y las otras sentencias de bloque no. Si ```test_expression_1``` es False pero ```test_expression_2``` es True, entonces ```block2_statement2 (s)``` se ejecutará y los demás no. Finalmente, si ```test_expression_1``` y ``` test_expression_2``` son ambos falsos, entonces se ejecutará el ```block3_statement (s)``` de la sección else. Esta estructura condicional de una declaración if permite controlar el flujo del código Python.

Usemos esto para filtrar los países que solo tienen una estimación de éxito.

# Ejercicio 9
Usaremos la declaración if anterior para eliminar los países con menos de una estimación de éxito. Para facilitar la visualización del resultado, almacenaremos las estimaciones medias de cada país en un nuevo diccionario ```country_means```.

In [251]:
def lista_nueva(lista):
  country_means={}
  for i in lista:
    if len(lista[i]) > 1:
      media = round(sum(lista[i])/len(lista[i]),2)
      country_means[i]=media
  return country_means

lista_nueva(success_estimates)



{'Australia': 0.3,
 'Brazil': 0.29,
 'Canada': 0.28,
 'France': 0.66,
 'Germany': 0.43,
 'Greece': 0.61,
 'Jamaica': 0.66,
 'Switzerland': 0.79,
 'Tunisia': 0.62,
 'USA': 0.33}

Al observar las medias de países resultantes, notamos que el país con la mayor estimación de éxito medio es Suiza con 0,79, mientras que la estimación de éxito medio más baja es Canadá con 0,28.

# Ejercicio 10
Después de revisar la política de la empresa sobre procedimientos estadísticos, observa que la empresa recomienda que todas las estimaciones (promedios, mínimos, máximos) deben tener al menos tres valores que contribuyan a la estadística de resumen. Escriba un ciclo for y use la estructura de la declaración if para seleccionar e imprimir las estimaciones de éxito promedio para los países que cumplen con esta política. Si el país no cumple con la política, escriba el nombre del país y "*No cumple con la política de la empresa*". Cada país debería aparecer en una nueva línea.

In [279]:
def promedios(lista_1):
  for i in lista_1:
    if len(lista_1[i]) > 2:
      prom = round(sum(lista_1[i])/len(lista_1[i]),2)
      print(f'País: {i}, estimación promedio:{prom}, mínimo: {min(lista_1[i])}, máximo: {max(lista_1[i])}')
     

promedios(success_estimates)


País: Australia, estimación promedio:0.3, mínimo: 0.11, máximo: 0.6
País: France, estimación promedio:0.66, mínimo: 0.2, máximo: 0.98
País: Brazil, estimación promedio:0.29, mínimo: 0.22, máximo: 0.43
País: USA, estimación promedio:0.33, mínimo: 0.2, máximo: 0.5
País: Greece, estimación promedio:0.61, mínimo: 0.15, máximo: 0.99
País: Jamaica, estimación promedio:0.66, mínimo: 0.61, máximo: 0.71
País: Switzerland, estimación promedio:0.79, mínimo: 0.51, máximo: 0.99
País: Germany, estimación promedio:0.43, mínimo: 0.36, máximo: 0.49


# Ejercicio 11
¿Cuál es otro enfoque para mejorar el problema de una muestra en algunos países? Piense en términos de los factores que impulsan la confianza en las decisiones comerciales basadas en datos.

(a) Agrupe los países en regiones más grandes para asegurarse de que cada región tenga al menos una estimación

(b) Solo elimine un país si sus estimaciones son muy grandes o muy pequeñas en comparación con otras estimaciones.

(c) Utilice una estadística de resumen diferente para el análisis que no sea el valor promedio.

(d) Revise por qué algunos países solo tienen una estimación y vea si se pueden obtener más datos para estos países.



### (d) Revise por qué algunos países solo tienen una estimación y vea si se pueden obtener más datos para estos países.


Hemos utilizado bucles y estructuras de control para calcular estadísticas resumidas parciales para cada uno de los países. Pongamos todo junto para obtener una recomendación sobre qué país deberíamos elegir para ampliar los servicios de vuelos de lujo.

# Ejercicio 12
Escriba el código para imprimir el nombre de cada país y las estadísticas resumidas. Cada línea debe mostrar un país y las estadísticas resumidas correspondientes: Estimación mínima (flotante), Estimación promedio (flotante), Estimación máxima (flotante), Número de estimaciones (int), Cumple con la política de la empresa de al menos 3 estimaciones (bool). Por ejemplo, la línea para Francia aparecería como:

```
Country: France , Min: 0.2 , Average: 0.655 , Max: 0.98 , NumEst: 4 , MeetsPolicy: True
```

In [291]:
def analisis(lista2):
  for i in lista2:
    min_= min(lista2[i])
    prom = round(sum(lista2[i])/len(lista2[i]),2)
    max_= max(lista2[i])
    long_= len(lista2[i])
    politica = long_ > 2

    print(f'Country: {i}, Min: {min_}, Average: {prom}, Max: {max_}, NumEst: {long_}, MeetsPolicy: {politica}')
    print('--------')

analisis(success_estimates)

Country: Australia, Min: 0.11, Average: 0.3, Max: 0.6, NumEst: 4, MeetsPolicy: True
--------
Country: France, Min: 0.2, Average: 0.66, Max: 0.98, NumEst: 4, MeetsPolicy: True
--------
Country: Italy, Min: 0.6, Average: 0.6, Max: 0.6, NumEst: 1, MeetsPolicy: False
--------
Country: Brazil, Min: 0.22, Average: 0.29, Max: 0.43, NumEst: 3, MeetsPolicy: True
--------
Country: USA, Min: 0.2, Average: 0.33, Max: 0.5, NumEst: 3, MeetsPolicy: True
--------
Country: England, Min: 0.45, Average: 0.45, Max: 0.45, NumEst: 1, MeetsPolicy: False
--------
Country: Canada, Min: 0.25, Average: 0.28, Max: 0.3, NumEst: 2, MeetsPolicy: False
--------
Country: Argentina, Min: 0.22, Average: 0.22, Max: 0.22, NumEst: 1, MeetsPolicy: False
--------
Country: Greece, Min: 0.15, Average: 0.61, Max: 0.99, NumEst: 6, MeetsPolicy: True
--------
Country: Morocco, Min: 0.29, Average: 0.29, Max: 0.29, NumEst: 1, MeetsPolicy: False
--------
Country: Tunisia, Min: 0.56, Average: 0.62, Max: 0.68, NumEst: 2, MeetsPolicy: F

# Conclusiones

Escribir_________________